Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift

```diff
@@ -0,0 +1,90 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+
+/// A sequence that groups its elements to optimize accesses.
+///
+/// You use a grouped sequence when you need efficient access to its elements via a grouping mechanism of your
+/// choosing and you aren't concerned with the order of the elements in the sequence.
+///
+/// ```swift
+/// var groupedSequence = GroupedSequence<Int, String> { $0.count }
+/// groupedSequence.append("a")
+/// groupedSequence.append(contentsOf: ["aa", "aaa"])
+///
+/// print(groupedSequence[1])
+/// // Prints ["a"]
+///
+/// print(groupedSequence[2])
+/// // Prints ["aa"]
+/// ```
+///
+/// You can iterate through a grouped sequence's unordered elements.
+///
+/// ```swift
+/// for item in groupedSequence {
+///     print(item)
+/// }
+///
+/// // Prints "aa"
+/// // Prints "a"
+/// // Prints "aaa"
+/// ```
+struct GroupedSequence<Key: Hashable, Element>: Sequence, CustomStringConvertible {
```

Comment:
Not needed for this PR but this already fulfills the requirements for `Collection` might be nice since we are making this a standalone data structure.

Reply:
hm, yes, looks like we're close! But we need to implement a few more requirements, which I think we should implement when we get a use for them. This isn't intended to be a general-purpose API—it's just for DocC's purposes.